### DISCUSSION TITLE
Why can't deduced return type be omitted here?

### DISCUSSION THREAD

**Discussion by jamadagni** [hsutter/cppfront#1247] [Q&A]
```
Int : type = {
    x : int;
    
    operator= : (out this, id: int) = { x = id; }
    
    operator+ : (this, val: int) -> _ = {
        return : Int = x + val;
    }

    value : (this) x;
}

main: () = {
    a : Int = 1;
    b := a + 2;
    std::cout << b.value() << std::endl;
}
```

If I omit the `-> _` from the `operator+`, I get the error:

    (7,9): error: return statement with expression must be in a function with a single anonymous return value

Why is this?

Two additional questions:

1) I hope `: Int = x + val` is the correct CPP2 equivalent of `Int(x + val)` from CPP1?
2) I hope this is the correct place to be asking questions to learn CPP2?

**Comment by jamadagni** [comment]
This seems related:

```operator== : (this, that) -> bool = id_ == that.id_ ;```

I am not allowed to omit `-> bool =` which gives me:

    error: operator== must return bool

I think this should be enforced as “operator== should return a boolean” rather than “operator== should be followed by `-> bool`”

**Comment by MaxSagebaum** [comment]
If you just omit the ` -> _` from the definition it currently defaults to no return type. If you currently also remove the `=` then it works fine. https://cpp2.godbolt.org/z/T7McKo1cK

The error for `operator==` seems to be regression. Also removing the `=` sign gives the same error. Do you want to create a bug report for it?

> I hope : Int = x + val is the correct CPP2 equivalent of Int(x + val) from CPP1?

This is ok, but `Int(x+val)` does also work.

**Reply by MaxSagebaum** [reply]
Missed the links:
 - Shorthand syntax: https://hsutter.github.io/cppfront/cpp2/functions/#generality-note-summary-of-function-defaults
 - Discussion about changes: https://github.com/hsutter/cppfront/discussions/714#discussioncomment-10268504
